// @coachcare/datepicker/esm5/datepicker-intl.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
/**
 * @record
 */
export function MatDatepickerIntlCatalog() { }
if (false) {
    /** @type {?} */
    MatDatepickerIntlCatalog.prototype.calendarLabel;
    /** @type {?} */
    MatDatepickerIntlCatalog.prototype.openCalendarLabel;
    /** @type {?} */
    MatDatepickerIntlCatalog.prototype.prevMonthLabel;
    /** @type {?} */
    MatDatepickerIntlCatalog.prototype.nextMonthLabel;
    /** @type {?} */
    MatDatepickerIntlCatalog.prototype.prevYearLabel;
    /** @type {?} */
    MatDatepickerIntlCatalog.prototype.nextYearLabel;
    /** @type {?} */
    MatDatepickerIntlCatalog.prototype.setToAMLabel;
    /** @type {?} */
    MatDatepickerIntlCatalog.prototype.setToPMLabel;
    /** @type {?} */
    MatDatepickerIntlCatalog.prototype.switchToMinuteViewLabel;
    /** @type {?} */
    MatDatepickerIntlCatalog.prototype.switchToHourViewLabel;
    /** @type {?} */
    MatDatepickerIntlCatalog.prototype.switchToMonthViewLabel;
    /** @type {?} */
    MatDatepickerIntlCatalog.prototype.switchToYearViewLabel;
    /** @type {?} */
    MatDatepickerIntlCatalog.prototype.switchToYearsViewLabel;
    /** @type {?} */
    MatDatepickerIntlCatalog.prototype.buttonSubmitText;
    /** @type {?} */
    MatDatepickerIntlCatalog.prototype.buttonSubmitLabel;
    /** @type {?} */
    MatDatepickerIntlCatalog.prototype.buttonCancelText;
    /** @type {?} */
    MatDatepickerIntlCatalog.prototype.buttonCancelLabel;
}
/**
 * Datepicker data that requires internationalization.
 */
var MatDatepickerIntl = /** @class */ (function () {
    /** Datepicker data that requires internationalization. */
    function MatDatepickerIntl() {
        /**
         * Stream that emits whenever the labels here are changed. Use this to notify
         * components if the labels have changed after initialization.
         */
        this.changes = new Subject();
        /** A label for the calendar popup (used by screen readers). */
        this.calendarLabel = 'Calendar';
        /** A label for the button used to open the calendar popup (used by screen readers). */
        this.openCalendarLabel = 'Open calendar';
        /** A label for the previous month button (used by screen readers). */
        this.prevMonthLabel = 'Previous month';
        /** A label for the next month button (used by screen readers). */
        this.nextMonthLabel = 'Next month';
        /** A label for the previous year button (used by screen readers). */
        this.prevYearLabel = 'Previous year';
        /** A label for the next year button (used by screen readers). */
        this.nextYearLabel = 'Next year';
        /** A label for the 'AM' button (used by screen readers). */
        this.setToAMLabel = 'Set date to AM';
        /** A label for the 'PM' button (used by screen readers). */
        this.setToPMLabel = 'Set date to PM';
        /** A label for the 'switch to minute view' button (used by screen readers). */
        this.switchToMinuteViewLabel = 'Change to minute view';
        /** A label for the 'switch to hour view' button (used by screen readers). */
        this.switchToHourViewLabel = 'Change to hour view';
        /** A label for the 'switch to month view' button (used by screen readers). */
        this.switchToMonthViewLabel = 'Change to month view';
        /** A label for the 'switch to year view' button (used by screen readers). */
        this.switchToYearViewLabel = 'Change to year view';
        /** A label for the 'switch to years view' button (used by screen readers). */
        this.switchToYearsViewLabel = 'Change to years view';
        /** Text for the 'submit' button. */
        this.buttonSubmitText = 'Ok';
        /** A label for the 'submit' button (used by screen readers). */
        this.buttonSubmitLabel = 'Choose the current date';
        /** Text for the 'cancel' button. */
        this.buttonCancelText = 'Cancel';
        /** A label for the 'cancel' button (used by screen readers). */
        this.buttonCancelLabel = 'Cancel the date selection';
    }
    MatDatepickerIntl.decorators = [
        { type: Injectable, args: [{ providedIn: 'root' },] },
    ];
    /** @nocollapse */ MatDatepickerIntl.ngInjectableDef = i0.defineInjectable({ factory: function MatDatepickerIntl_Factory() { return new MatDatepickerIntl(); }, token: MatDatepickerIntl, providedIn: "root" });
    return MatDatepickerIntl;
}());
export { MatDatepickerIntl };
if (false) {
    /**
     * Stream that emits whenever the labels here are changed. Use this to notify
     * components if the labels have changed after initialization.
     * @type {?}
     */
    MatDatepickerIntl.prototype.changes;
    /**
     * A label for the calendar popup (used by screen readers).
     * @type {?}
     */
    MatDatepickerIntl.prototype.calendarLabel;
    /**
     * A label for the button used to open the calendar popup (used by screen readers).
     * @type {?}
     */
    MatDatepickerIntl.prototype.openCalendarLabel;
    /**
     * A label for the previous month button (used by screen readers).
     * @type {?}
     */
    MatDatepickerIntl.prototype.prevMonthLabel;
    /**
     * A label for the next month button (used by screen readers).
     * @type {?}
     */
    MatDatepickerIntl.prototype.nextMonthLabel;
    /**
     * A label for the previous year button (used by screen readers).
     * @type {?}
     */
    MatDatepickerIntl.prototype.prevYearLabel;
    /**
     * A label for the next year button (used by screen readers).
     * @type {?}
     */
    MatDatepickerIntl.prototype.nextYearLabel;
    /**
     * A label for the 'AM' button (used by screen readers).
     * @type {?}
     */
    MatDatepickerIntl.prototype.setToAMLabel;
    /**
     * A label for the 'PM' button (used by screen readers).
     * @type {?}
     */
    MatDatepickerIntl.prototype.setToPMLabel;
    /**
     * A label for the 'switch to minute view' button (used by screen readers).
     * @type {?}
     */
    MatDatepickerIntl.prototype.switchToMinuteViewLabel;
    /**
     * A label for the 'switch to hour view' button (used by screen readers).
     * @type {?}
     */
    MatDatepickerIntl.prototype.switchToHourViewLabel;
    /**
     * A label for the 'switch to month view' button (used by screen readers).
     * @type {?}
     */
    MatDatepickerIntl.prototype.switchToMonthViewLabel;
    /**
     * A label for the 'switch to year view' button (used by screen readers).
     * @type {?}
     */
    MatDatepickerIntl.prototype.switchToYearViewLabel;
    /**
     * A label for the 'switch to years view' button (used by screen readers).
     * @type {?}
     */
    MatDatepickerIntl.prototype.switchToYearsViewLabel;
    /**
     * Text for the 'submit' button.
     * @type {?}
     */
    MatDatepickerIntl.prototype.buttonSubmitText;
    /**
     * A label for the 'submit' button (used by screen readers).
     * @type {?}
     */
    MatDatepickerIntl.prototype.buttonSubmitLabel;
    /**
     * Text for the 'cancel' button.
     * @type {?}
     */
    MatDatepickerIntl.prototype.buttonCancelText;
    /**
     * A label for the 'cancel' button (used by screen readers).
     * @type {?}
     */
    MatDatepickerIntl.prototype.buttonCancelLabel;
}
